Source/WebCore/inspector/InspectorStyleSheet.cpp

```diff
@@ -1535,22 +1504,61 @@ ExceptionOr<void> InspectorStyleSheet::setRuleStyleText(const InspectorCSSId& id
         ASSERT(childStart <= childEnd);
         RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(childEnd <= styleSheetText.length());
 
-        replacementBodyText.append('\n', indentation);
+        canonicalRuleText.append('\n', indentation);
 
         // Non-style rules don't include the `@rule` prelude in their header range.
         if (childRuleSourceData->type != StyleRuleType::Style)
-            replacementBodyText.append(atRuleIdentifierForType(childRuleSourceData->type));
+            canonicalRuleText.append(atRuleIdentifierForType(childRuleSourceData->type));
 
-        replacementBodyText.appendSubstring(styleSheetText, childStart, childEnd - childStart);
-        replacementBodyText.append("}\n"_s);
+        canonicalRuleText.appendSubstring(styleSheetText, childStart, childEnd - childStart);
+        canonicalRuleText.append("}\n"_s);
     }
 
-    auto closingIndentationLineStart = newText.reverseFind('\n');
+    auto closingIndentationLineStart = ruleStyleDeclarationText.reverseFind('\n');
     if (closingIndentationLineStart != notFound)
-        replacementBodyText.appendSubstring(newText, closingIndentationLineStart);
+        canonicalRuleText.appendSubstring(ruleStyleDeclarationText, closingIndentationLineStart);
 
-    auto patchedStyleSheetText = makeStringByReplacing(styleSheetText, bodyStart, bodyEnd - bodyStart , replacementBodyText);
-    setPatchedStyleSheetText(patchedStyleSheetText);
+    return canonicalRuleText;
+}
+
+ExceptionOr<void> InspectorStyleSheet::setRuleStyleText(const InspectorCSSId& id, const String& newStyleDeclarationText, String* oldStyleDeclarationText, const String* newRuleText, String* oldRuleText)
+{
+    auto* cssRule = ruleForId(id);
+    if (!cssRule)
+        return Exception { ExceptionCode::NotFoundError };
+
+    RefPtr<CSSRuleSourceData> sourceData = ruleSourceDataFor(cssRule);
+    if (!sourceData)
+        return Exception { ExceptionCode::NotFoundError };
+
+    RefPtr<CSSRuleSourceData> logicalContainingRuleSourceData = sourceData->isImplicitlyNested ? ruleSourceDataFor(cssRule->parentRule()) : sourceData;
+    if (!logicalContainingRuleSourceData)
+        return Exception { ExceptionCode::NotFoundError };
+
+    auto* cssStyleDeclaration = styleForId(id);
+    if (!cssStyleDeclaration)
+        return Exception { ExceptionCode::NotFoundError };
+
+    if (oldStyleDeclarationText)
+        *oldStyleDeclarationText = cssStyleDeclaration->cssText();
+
+    cssStyleDeclaration->setCssText(newStyleDeclarationText);
+
+    unsigned bodyStart = logicalContainingRuleSourceData->ruleBodyRange.start;
+    unsigned bodyEnd = logicalContainingRuleSourceData->ruleBodyRange.end;
+    ASSERT(bodyStart <= bodyEnd);
+
+    const String& styleSheetText = m_parsedStyleSheet->text();
+    RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(bodyEnd <= styleSheetText.length());
+
+    if (oldRuleText)
+        *oldRuleText = styleSheetText.substring(bodyStart, bodyEnd - bodyStart);
```

Comment:
should this happen before we call `CSSStyleDeclaration::setCssText`?  i.e. grab the value for both `old*Text` before modifying anything